pylint/checkers/variables.py

```diff
@@ -979,267 +1001,296 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
-                                    # Doing the following is fine:
-                                    #   class A:
-                                    #      x = 42
-                                    #      y = lambda attr=x: attr
-                                    self.add_message(
-                                        "used-before-assignment",
-                                        args=node.name,
-                                        node=node,
-                                    )
-                            else:
-                                self.add_message(
-                                    "undefined-variable", args=node.name, node=node
-                                )
-                        elif current_consumer.scope_type == "lambda":
-                            self.add_message(
-                                "undefined-variable", node=node, args=node.name
-                            )
-                elif self._is_only_type_assignment(node, defstmt):
-                    self.add_message("undefined-variable", node=node, args=node.name)
-                elif isinstance(defstmt, nodes.ClassDef):
-                    is_first_level_ref = self._is_first_level_self_reference(
-                        node, defstmt
-                    )
-                    if is_first_level_ref == 2:
                         self.add_message(
-                            "used-before-assignment", node=node, args=node.name
+                            "undefined-variable", args=node.name, node=node
                         )
-                    if is_first_level_ref:
-                        break
-                elif isinstance(defnode, nodes.NamedExpr):
-                    if isinstance(defnode.parent, nodes.IfExp):
-                        if self._is_never_evaluated(defnode, defnode.parent):
-                            self.add_message(
-                                "undefined-variable", node=node, args=node.name
-                            )
+                        return (VariableVisitConsumerAction.CONSUME, found_nodes)
 
-            current_consumer.mark_as_consumed(node.name, found_nodes)
-            # check it's not a loop variable used outside the loop
-            self._loopvar_name(node)
-            break
-        else:
-            # we have not found the name, if it isn't a builtin, that's an
-            # undefined name !
-            if undefined_variable_is_enabled and not (
-                node.name in nodes.Module.scope_attrs
-                or utils.is_builtin(node.name)
-                or node.name in self.config.additional_builtins
-                or (
-                    node.name == "__class__"
-                    and isinstance(frame, nodes.FunctionDef)
-                    and frame.is_method()
-                )
-            ):
-                if not utils.node_ignores_exception(node, NameError):
+            if base_scope_type != "lambda":
```

Comment:
```suggestion
            elif base_scope_type != "lambda":
```
This needs to be an `elif` as the `return` above only captures the inner most conditional.

Reply:
Do you know if this if statement makes sense any way? (Also look at comment below this).
Using `if`, `elif` and even removing the check altogether all pass the tests. I'm (obviously) fine with the `elif` and indenting the other comment, but it seems we're missing something in our tests here. Did you, by any chance, find out what during your review?

Reply:
I wasn't able to come up with an example on the fly. Might be that we get lucky here.
IMO it doesn't hurt for check conditions to be as precise as possible, so I would recommend to leave it as it was. That seemed to work.

Furthermore, using `elif` just makes sure we don't perform unnecessary work and thus could save some time.